pilot/pkg/serviceregistry/kube/controller/controller.go

```diff
@@ -655,6 +655,39 @@ func (c *Controller) getEndpoints(ip string, endpointPort int32, svcPort *model.
 	}
 }
 
+// getServiceInstanceByEndpoint returns the service instances represented by the given endpoints.
+func (c *Controller) getServiceInstanceByEndpoint(endpoints v1.Endpoints) []*model.ServiceInstance {
+	out := make([]*model.ServiceInstance, 0)
+
+	hostname := kube.ServiceHostname(endpoints.Name, endpoints.Namespace, c.domainSuffix)
+	c.RLock()
+	svc := c.servicesMap[hostname]
+	c.RUnlock()
+
+	if svc != nil {
+		for _, ss := range endpoints.Subsets {
+			for _, port := range ss.Ports {
```

Comment:
do we need all ip:port or just a all unique IPs? I don't know if there are cases where the distinction matters. Maybe we keep it like this to be safe, as I doubt it has tangible performance impacts

Reply:
For our specific case, we only need the IPs, but it seems correct for `Controller.AppendInstanceHandler()` to invoke the callback for every ServiceInstance (==> ip:port). Given that the penalty is (supposedly) low (see [comment](https://github.com/istio/istio/pull/15950#discussion_r314970646)), I think we can just leave it like that.